crates/services/txpool/src/service/update_sender.rs

```diff
@@ -0,0 +1,362 @@
+
+impl<P, Tx> SendStatus for Sender<P, Tx>
+where
+    Tx: SendStatus,
+{
+    fn try_send(&mut self, msg: TxStatusMessage) -> Result<(), SendError> {
+        // Add the message to the stream.
+        self.stream.add_msg(msg);
+
+        // Try to send the next message in the stream.
+        if let Some(msg) = self.stream.try_next() {
+            // Send the message to the outgoing channel.
+            match self.tx.try_send(msg) {
+                Ok(()) => (),
+                // If the channel is full, add a failure to the stream.
+                Err(SendError::Full) => self.stream.add_failure(),
+                // If the channel is closed, close the stream.
+                Err(SendError::Closed) => self.stream.close_recv(),
+            }
+        }
+
+        // Check if the stream is now closed.
+        if self.stream.is_closed() {
+            Err(SendError::Closed)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn is_closed(&self) -> bool {
+        self.stream.is_closed()
+    }
+
+    fn is_full(&self) -> bool {
+        self.tx.is_full()
+    }
+}
+
+impl SendStatus for mpsc::Sender<TxStatusMessage> {
+    fn try_send(&mut self, msg: TxStatusMessage) -> Result<(), SendError> {
+        match (*self).try_send(msg) {
+            Ok(()) => Ok(()),
+            Err(TrySendError::Full(_)) => Err(SendError::Full),
+            Err(TrySendError::Closed(_)) => Err(SendError::Closed),
+        }
+    }
+
+    fn is_closed(&self) -> bool {
+        self.is_closed()
+    }
+
+    fn is_full(&self) -> bool {
+        self.capacity() == 0
+    }
+}
+
+impl UpdateSender {
+    /// Create a new UpdateSender with a specified capacity for the semaphore
+    pub fn new(capacity: usize) -> UpdateSender {
+        UpdateSender {
+            senders: Default::default(),
+            permits: Arc::new(Semaphore::new(capacity)),
+        }
+    }
+
+    /// Try to subscribe for updates, returns a TxStatusStream if successful
+    #[cfg(test)]
+    pub fn try_subscribe<C>(&self, tx_id: Bytes32) -> Option<TxStatusStream>
+    where
+        C: CreateChannel,
+    {
+        // Remove closed senders from the list
+        remove_closed(&mut self.senders.lock());
+
+        // Try to acquire a permit from the semaphore
+        let permit = Arc::clone(&self.permits).try_acquire()?;
+
+        // Call subscribe_inner with the acquired permit
+        Some(self.subscribe_inner::<C>(tx_id, permit))
+    }
+
+    /// Asynchronously await a permit to subscribe
+    /// returns a TxStatusStream on completion
+    pub async fn subscribe<C>(&self, tx_id: Bytes32) -> TxStatusStream
+    where
+        C: CreateChannel,
+    {
+        // Remove closed senders from the list.
+        // Careful not to hold the lock while awaiting.
+        {
+            remove_closed(&mut self.senders.lock());
+        }
+
+        // Acquire a permit from the semaphore asynchronously.
+        let permit = Arc::clone(&self.permits).acquire().await;
```

Comment:
Could this cause messages to get missed if the creation of the sender is awaiting a permit?

Reply:
are you suggesting a message wouldn't be sent in this scenario?

Reply:
or that the transaction could complete before it is possible to finalize the subscription?

Reply:
we could require subscribing before the transaction is submitted given that the id is known but this could be abused. another option would be holding an LRU of recent tx that can still be subscribed for, but that supposes timing requirements

Reply:
Yes it will cause messages to be missed but we can't garauntee any sort of message delivery anyway